src/routes/getFiles.ts

```diff
@@ -0,0 +1,258 @@
+const asGetFilesBody = asObject({
+  repoId: asNonEmptyString,
+  ignoreTimestamps: asOptional(asBoolean),
+  paths: asMap(asNumber)
+})
+
+interface GetFilesResponseData {
+  total: number
+  paths: GetFilesResponsePaths
+}
+
+interface GetFilesResponsePaths {
+  [path: string]: GetFilesResponseFile | GetFilesResponseDirectory
+}
+
+interface GetFilesResponseFile extends StoreFile {
+  timestamp: number
+}
+
+interface GetFilesResponseDirectory {
+  paths: StoreFileMap
+  timestamp: number
+}
+
+export const getFilesRouter = Router()
+
+getFilesRouter.post('/getFiles', async (req, res) => {
+  let body: GetFilesBody
+
+  // Validate request body
+  try {
+    body = asGetFilesBody(req.body)
+  } catch (err) {
+    throw makeApiClientError(400, err.message)
+  }
+
+  const { repoId, ignoreTimestamps = false } = body
+  const paths = Object.keys(body.paths)
+
+  // Use max page size config to limit the paths processed
+  if (paths.length > maxPageSize) {
+    throw makeApiClientError(
+      422,
+      `Too many paths. maxPageSize is ${maxPageSize}`
+    )
+  }
+
+  // Apart of the repsonse
+  const responsePaths: GetFilesResponsePaths = {}
+
+  // Document keys to fetch
+  const documentKeys: string[] = []
+
+  // Map for preparing parent documents query
+  const documentKeysMapOfParentDocumentKeys: {
+    [parentDocumentKey: string]: string[]
+  } = {}
+  // Map for timestamps from parent documents
+  const documentTimestampMapOfDocumentKeys: {
+    [documentKey: string]: number
+  } = {}
+
+  // Prepare document parents query
+  paths.forEach(documentPath => {
+    const documentKey = `${repoId}:${documentPath}`
+    const parentPaths = getParentPathsOfPath(documentPath)
+
+    // For repo documents, no need to check parent document, so directly the
+    // repo document key it into the documentKeys for later retrieval.
+    if (documentPath === '/') {
+      documentKeys.push(`${repoId}:${documentPath}`)
+    } else {
+      const parentPath = parentPaths.length > 0 ? parentPaths[0] : '/'
+      const parentKey = `${repoId}:${parentPath}`
+
+      documentKeysMapOfParentDocumentKeys[parentKey] =
+        documentKeysMapOfParentDocumentKeys[parentKey] === undefined
+          ? [documentKey]
+          : documentKeysMapOfParentDocumentKeys[parentKey].concat(documentKey)
+    }
+  })
+
+  // Fetch Document Parents:
+
+  const parentKeys = Object.keys(documentKeysMapOfParentDocumentKeys)
+  if (parentKeys.length > 0) {
+    const fetchParentDocumentsResults = await dataStore.fetch({
+      keys: parentKeys
+    })
+    for (const row of fetchParentDocumentsResults.rows) {
+      const parentDocumentKey = row.key
+      if ('doc' in row) {
+        let parentDocument: StoreDirectoryDocument
+
+        try {
+          // For directory document
+          parentDocument = asStoreDirectoryDocument(row.doc)
+        } catch (err) {
+          throw makeApiClientError(422, `File '${row.key}' is not a directory.`)
```

Comment:
Throw 500 instead of 455 here (new Error). There's no way this would happen unless the database is currupt.